planner/core/plan_cache_lru.go

```diff
@@ -0,0 +1,176 @@
+	}
+	return &LRUPlanCache{
+		capacity:       capacity,
+		size:           0,
+		buckets:        make(map[string]map[*list.Element]struct{}, 1), //Generally one query has one plan
+		lruList:        list.New(),
+		pickFromBucket: pickFromBucket,
+	}, nil
+}
+
+// Get tries to find the corresponding value according to the given key.
+func (l *LRUPlanCache) Get(key kvcache.Key, paramTypes []*types.FieldType) (value kvcache.Value, ok bool) {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	bucket, bucketExist := l.buckets[string(hack.String(key.Hash()))]
+	if bucketExist {
+		if element, exist := l.pickFromBucket(bucket, paramTypes); exist {
+			l.lruList.MoveToFront(element)
+			return element.Value.(*planCacheEntry).PlanValue, true
+		}
+	}
+	return nil, false
+}
+
+// Put puts the (key, value) pair into the LRU Cache.
+func (l *LRUPlanCache) Put(key kvcache.Key, value kvcache.Value, paramTypes []*types.FieldType) {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	hash := string(key.Hash())
+	bucket, bucketExist := l.buckets[hash]
+	// bucket exist
+	if bucketExist {
+		if element, exist := l.pickFromBucket(bucket, paramTypes); exist {
+			element.Value.(*planCacheEntry).PlanValue = value
+			l.lruList.MoveToFront(element)
+			return
+		}
+	} else {
+		l.buckets[hash] = make(map[*list.Element]struct{}, 1)
+	}
+
+	newCacheEntry := &planCacheEntry{
+		PlanKey:   key,
+		PlanValue: value,
+	}
+	element := l.lruList.PushFront(newCacheEntry)
+	l.buckets[hash][element] = struct{}{}
+	l.size++
+	if l.size > l.capacity {
+		l.removeOldest()
+	}
+}
+
+// Delete deletes the multi-values from the LRU Cache.
+func (l *LRUPlanCache) Delete(key kvcache.Key) {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	hash := hack.String(key.Hash())
+	bucket, bucketExist := l.buckets[string(hash)]
+	if bucketExist {
+		for element := range bucket {
+			l.lruList.Remove(element)
+			l.size--
+		}
+		delete(l.buckets, string(hash))
+	}
+}
+
+// DeleteAll deletes all elements from the LRU Cache.
+func (l *LRUPlanCache) DeleteAll() {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	for lru := l.lruList.Back(); lru != nil; lru = l.lruList.Back() {
+		l.lruList.Remove(lru)
+		l.size--
+	}
+	l.buckets = make(map[string]map[*list.Element]struct{}, 1)
+}
+
+// Size gets the current cache size.
+func (l *LRUPlanCache) Size() int {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	return int(l.size)
+}
+
+// SetCapacity sets capacity of the cache.
+func (l *LRUPlanCache) SetCapacity(capacity uint) error {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	if capacity < 1 {
+		return errors.New("capacity of lru cache should be at least 1")
+	}
+	l.capacity = capacity
+	for l.size > l.capacity {
+		l.removeOldest()
+	}
+	return nil
+}
+
+// removeOldest removes the oldest element from the cache.
+func (l *LRUPlanCache) removeOldest() {
+	lru := l.lruList.Back()
+	if l.onEvict != nil {
+		l.onEvict(lru.Value.(*planCacheEntry).PlanKey, lru.Value.(*planCacheEntry).PlanValue)
+	}
```

Comment:
What does this part mean?

Reply:
> What does this part mean?

'onEvict' will be called if any eviction happened, it is only used for test  now